### PR TITLE
#141 unicode + case-insensitive configuration bug

### PR DESCRIPTION
**Pull request by gdziadkiewicz**
Depends on #157 
Fixes part of #141

### Description:
As described in #141 currently case-insensitivity does not work together with Unicode due to a bug in handling Unicode Categories. This PR adds tests for that and fixes the bug by adding a branch that handles Unicode Categories. It leaves nonletter categories unmodified and transforms letter categories to include all cases.

### CODE REVIEW

File: src/FsLex.Core/fslexast.fs [hunk 1]
```diff
@@ -64,14 +60,23 @@ let NumUnicodeCategories = unicodeCategories.Count
 assert (NumUnicodeCategories = 30) // see table interpreter
 let encodedUnicodeCategoryBase = 0xFFFFFF00u
 
-let EncodeUnicodeCategoryIndex(idx:int) = encodedUnicodeCategoryBase + uint32 idx
-let EncodeUnicodeCategory s: Parser<uint32> = fun ctx ->
+let EncodeUnicodeCategoryIndex(idx:int) : Alphabet = encodedUnicodeCategoryBase + uint32 idx
+let EncodeUnicodeCategory(s:string) : Parser<Alphabet> = fun ctx ->
```
Comment (gdziadkiewicz) [inline src/FsLex.Core/fslexast.fs:64]: I will revert this as it is not connected to the PR.

File: src/FsLex.Core/fslexast.fs [hunk 2]
```diff
@@ -64,14 +60,23 @@ let NumUnicodeCategories = unicodeCategories.Count
...
     if not ctx.unicode then
         failwith "unicode category classes may only be used if --unicode is specified"
-    if unicodeCategories.ContainsKey(s) then
-        EncodeUnicodeCategoryIndex (int32 unicodeCategories.[s])
+    match unicodeCategories.TryGetValue s with
+    | true,  v -> v |> int32 |> EncodeUnicodeCategoryIndex
+    | false, _ -> failwithf "invalid Unicode category: '%s'" s
+
+//TODO: Test
```
Comment (gdziadkiewicz) [inline src/FsLex.Core/fslexast.fs:71]: Add unit tests for that.

File: tests/fsyacc/OldFsYaccTests.fsx [hunk 3]
```diff
@@ -161,6 +163,17 @@ runTests test1unicodeProj [
     test1Input1, test1Input1Bsl
     sprintf "%s %s" test1Input2Variation1 test1Input2Variation2, test1Input2Bsl
     sprintf "--tokens %s" test1unicodeInput3, test1unicodeInput3TokensBsl
+    //sprintf "--tokens %s" test1unicodeInput4, test1unicodeInput4TokensBsl
```
Comment (gdziadkiewicz) [inline tests/fsyacc/OldFsYaccTests.fsx:166]: I would like to transform this line into a tests that shows that the lexer without `-i` fails on this input.

File: tests/fsyacc/unicode/test1-unicode.input4.utf8 [hunk 4]
```diff
@@ -0,0 +1,9 @@
+
+next line tests one unicode character class
+ÄËÖÏÜâæçñõö + id
+next line tests specific unicode characters
+≠ ≠≠ ≈≈ ≈≈≈
+id + id
+next line tests some more random unicode characters
+ǲМНОПРСТУФХЦẀẁẂќ αβΛΘΩΨΧΣδζȚŶǺ
```
Comment (gdziadkiewicz) [inline tests/fsyacc/unicode/test1-unicode.input4.utf8:8]: I will make it more visible that the whole test is about this `ǲ` that is a single title case letter.

File: tests/fsyacc/OldFsYaccTests.fsx (hunk 3 above)
Comment (gdziadkiewicz) [reply tests/fsyacc/OldFsYaccTests.fsx]: Done

File: tests/fsyacc/unicode/test1-unicode.input4.utf8 (hunk 4 above)
Comment (gdziadkiewicz) [reply tests/fsyacc/unicode/test1-unicode.input4.utf8]: Done, changed name.

File: src/FsLex.Core/fslexast.fs (hunk 1 above)
Comment (gdziadkiewicz) [reply src/FsLex.Core/fslexast.fs]: Reverted.

File: src/FsLex.Core/fslexast.fs [hunk 5]
```diff
@@ -68,10 +64,18 @@ let EncodeUnicodeCategoryIndex(idx:int) = encodedUnicodeCategoryBase + uint32 id
 let EncodeUnicodeCategory s: Parser<uint32> = fun ctx ->
     if not ctx.unicode then
         failwith "unicode category classes may only be used if --unicode is specified"
-    if unicodeCategories.ContainsKey(s) then
-        EncodeUnicodeCategoryIndex (int32 unicodeCategories.[s])
+    match unicodeCategories.TryGetValue s with
```
Comment (gdziadkiewicz) [inline src/FsLex.Core/fslexast.fs:67]: Rollback as it is a refactoring not connected to the issue.

File: src/FsLex.Core/fslexast.fs [hunk 6]
```diff
@@ -92,15 +97,15 @@ let EncodeChar(c:char): Parser<uint32> = fun ctx ->
                  specificUnicodeCharsDecode.[idx] <- c
              specificUnicodeChars.[c]
      else
-         if x >= 256u then failwithf "the Unicode character '%x' may not be used unless --unicode is specified" <| int c
+         if x >= 256u then failwithf "the Unicode character '0x%x' may not be used unless --unicode is specified" <| int c
```
Comment (gdziadkiewicz) [inline src/FsLex.Core/fslexast.fs:100]: Before: `FSLEX: error FSL000: the Unicode character '178' may not be used unless --unicode is specified`
After: `FSLEX: error FSL000: the Unicode character '0x178' may not be used unless --unicode is specified`

This resolves the ambiguity between https://www.compart.com/en/unicode/U+0178 and https://www.compart.com/en/unicode/U+00B2

File: src/FsLex.Core/fslexast.fs (hunk 5 above)
Comment (gdziadkiewicz) [reply src/FsLex.Core/fslexast.fs]: Done

**Comment by gdziadkiewicz** [comment]
This is ready for review.

**Review by dsyme — APPROVED** [review]

**Comment by dsyme** [comment]
Great work!

**Pull request merged**